rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,433 @@
+  forbidden, even if they overlap - to enable our use case, we’d have to verify
+  the existence of all the partitions we intend to delete, followed by doing all
+  the (SET, CLEAR) operations OR rollback / cancel the transaction.
+
+* While this strategy allows for serializing the transactions into a
+  conflict-free state, it does not help with each VAST node "seeing the full
+  picture", i.e., being aware of the transactions that are triggered from within
+  other VAST nodes.
+
+* There doesn’t seem to be a full CDC (change data capture) functionality, only
+  watching single keys - there could be a key for a "transaction_id" (an integer
+  increment for every transaction), and from there we could work around the
+  problem by requesting all recent transactions whenever this key is modified.
+
+We need to read further on how to XXX.
+
+##### etcd
+
+Etcd is a distributed key-value store that uses Raft for quorum-based consensus. From a CAP Theorem perspective, this brings us into the CP realm, because a master must be elected to perform writes. Stale reads may be supported, though, which we can tolerate.
+
+Here are some incomplete notes of taking a first look:
+
+- Multi write transactions are shaped like if-then-else statements, they always
+  have a comparison, a list of writes on success and a list of writes on failure
+
+- Watches are possible on single keys and prefixes in the key-space, which would
+  allow us to model our use-case naturally.
+
+- Etcd comes with a gRPC API, making it easy to embed for us. We're already
+  depending on gRPC transitively via Apache Arrow (cf. Flight).
+
+##### Consul
+
+TBD
+
+##### Native Raft
+
+Instead of depending on etcd, we may consider our own actor-based Raft
+implementation. In this case, Raft would be a frontend for a key-value store
+that supports the desired transactional operations natively.
+
+The shape of a metastore would be dedicated CAF service that VAST can speak
+natively to. Not every VAST node should include a Raft metastore, as a typical
+number of Raft nodes remains in the single digits. That is, multiple VAST nodes
+share a single metastore. We would need at least 3 Raft instances to obtain a
+quorum majority when 1 node fails, 5 Raft nodes to tolerate 2 node failures.
+
+VAST already had a
+[working-albeit-minimal](https://github.com/tenzir/vast/pull/926) implementation
+in the past that we removed, given that we didn't need the functionality at the
+time. There are some [open questions on composability of log and state
+machine](https://groups.google.com/g/raft-dev/c/O80OSWrQieo) that we should
+revisit.
+
+## Alternatives
+
+TBD
+
+## Annex
+
+### A1: Considerations regarding the persistence layer as a shared component
+
+As partitions are identified by UUID we consider the write end mostly
+conflict-free. In case a name clash happens anyways the filesystem abstraction
+can just reject one of the writes, this problem can be solved client side by
```

Comment:
I wrote this paragraph as an intro to the rest of these considerations. Removing it would make the rest harder to follow, even though the issue here is rather theoretical.